pkg/kv/kvserver/rangefeed/scheduler.go

```diff
@@ -357,41 +420,83 @@ func (s *Scheduler) processEvents(ctx context.Context) {
 // Any attempts to enqueue events for processor after this call will return an
 // error.
 func (s *Scheduler) Unregister(id int64) {
-	s.mu.Lock()
-	defer s.mu.Unlock()
+	shard := s.shards[shardIndex(id, len(s.shards))]
+	shard.Lock()
+	defer shard.Unlock()
 
-	delete(s.mu.procs, id)
-	delete(s.mu.status, id)
+	delete(shard.procs, id)
+	delete(shard.status, id)
 }
 
 func (s *Scheduler) Stop() {
-	// Stop all processors.
-	s.mu.Lock()
-	if !s.mu.quiescing {
-		// On first close attempt trigger termination of all unfinished callbacks,
-		// we only need to do that once to avoid closing s.drained channel multiple
-		// times.
-		s.mu.quiescing = true
+	// Stop all processors across all shards.
+	for _, shard := range s.shards {
+		shard.Lock()
+		if !shard.quiescing {
+			// On first close attempt trigger termination of all unfinished callbacks,
+			// we only need to do that once to avoid closing s.drained channel multiple
+			// times.
+			shard.quiescing = true
+		}
+		shard.Unlock()
+		shard.cond.Broadcast()
 	}
-	s.mu.Unlock()
-	s.cond.Broadcast()
 	s.wg.Wait()
 
 	// Synchronously notify all non-stopped processors about stop.
-	s.mu.Lock()
-	for id, p := range s.mu.procs {
-		pending := s.mu.status[id]
-		// Ignore processors that already processed their stopped event.
-		if pending == Stopped {
-			continue
+	for _, shard := range s.shards {
+		shard.Lock()
```

Comment:
IMHO division line between what scheduler does vs what shard does it arbitrary. Sometimes scheduler gets the lock and changes shard, sometimes its shard itself. Makes it harder to change. Quiescing and notifying processors could be part of shard.

Reply:
Moved all shard logic into shard methods.